app/lib/plugin_initializer.rb

```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+module DiscourseSolved
+  class PluginInitializer
+    attr_reader :plugin
+
+    def initialize(plugin)
+      @plugin = plugin
+    end
+
+    def apply_plugin_api
+      # this method should be overridden by subclasses
+      raise NotImplementedError
+    end
+  end
+
+  class AssignsRemainderForTopicsQuery < PluginInitializer
+    def apply_plugin_api
+      plugin.register_modifier(:assigns_reminder_assigned_topics_query) do |query|
+        next query if !SiteSetting.ignore_solved_topics_in_assigned_reminder
+        query.where(
+          "topics.id NOT IN (
+            SELECT topic_id
+            FROM topic_custom_fields
+            WHERE name = 'accepted_answer_post_id'
+          )",
+        )
+      end
```

Comment:
I wonder if that doesn't deserve a better API than a plugin modifier. This create a very strong relationship between assign code and this code.

Reply:
Ideally a good API doesn't require changes if the core code implementation changes, which is very unlikely here.

Reply:
I personally think this isn't an issue. The modifier is generic enough that other plugins may use it, it seems like a good API that may be re-used. Gabriel is adding tests in the application of the modifier so we will detect any breaking changes on the discourse-assign side. This is how I would implement it as well

Reply:
It's not about if it can be used by other plugins or not. My point is that if we change how the query is implemented in discourse-assign every plugin will have to change, which goes against the idea of an API. You build API so you don't expose core. So my suggestion is to reverse the API, so you can declare in a nice way what kind of filter you want to create and let discourse-assign how it will filter: sql, ruby, ...

Reply:
how would modifiers even work then if this is a problem? Every instance of apply_modifier is a version of this problem then, correct?

Reply:
To a certain extend yes. But 1) it's not like apply_modifier is our only solution
2) this case is kinda extreme because it's a query, if you were returning a list of results and filtering them that would be less annoying

This being said, Im not here to block your PR, Im just here to note that maybe we could do better here. If you want to go with this go with this 👍